KAFKA-8904: Improve producer's topic metadata fetching.

bdbyrne opened the pull request:
Improves the producer's topic metadata fetching logic by maintaining a per-topic last refresh time, which is used to discriminate which topics' metadata should be fetched.

The improvement can be witnessed during producer startup, where many topics may be fetched serially. Previously, where N topics were being fetched serially, O(N^2) total topic metadata would be processed, whereas this improvement makes it O(N).

A simple test can show the differences fairly easily. Starting a producer on 500 existing topics with 64 partitions each, single-threaded:

Old: 1000 records sent, 32.446463 records/sec (0.00 MB/sec), 65.52 ms avg latency, 459.00 ms max latency, 23 ms 50th, 213 ms 95th, 360 ms 99th, 459 ms 99.9th.

New: 1000 records sent, 206.143063 records/sec (0.00 MB/sec), 10.55 ms avg latency, 187.00 ms max latency, 7 ms 50th, 29 ms 95th, 43 ms 99th, 187 ms 99.9th.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

dengziming commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerMetadata.java:
```diff
@@ -72,16 +157,68 @@ public synchronized boolean containsTopic(String topic) {
+    @Override
+    public synchronized long computeTimeToNextUpdate(long nowMs) {
+        long nextUpdateMs = super.computeTimeToNextUpdate(nowMs);
+        if (nextUpdateMs == 0) {
```
should this `if (nextUpdateMs == 0) ` be `if (nextUpdateMs <= 0) {`

bdbyrne replied:
This would be safer, yes. Done.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/Metadata.java:
```diff
@@ -286,17 +300,27 @@ private void checkUnauthorizedTopics(Cluster cluster) {
                 continue;
 
+            if (metadata.isInternal())
+                internalTopics.add(metadata.topic());
```
This was inside the check for errors before but we have moved it out. Is that because it doesn't matter either way?

bdbyrne replied:
That's correct. I believe it was a bug to maintain a list of internal topics that we don't need to retain, but from my cursory check, the list of internal topics isn't really used anywhere in the producer or consumer.

bdbyrne replied:
Also note as part of this refactoring, TOPIC_AUTHORIZATION_FAILED and INVALID_TOPIC_EXCEPTION errors for retainTopic() = false are now being ignored too, whereas they weren't before.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/Metadata.java:
```diff
@@ -307,20 +331,33 @@ private MetadataCache handleMetadataResponse(MetadataResponse metadataResponse,
+                if (metadata.error() == Errors.INVALID_TOPIC_EXCEPTION) {
+                    invalidTopics.add(metadata.topic());
+                } else if (metadata.error() == Errors.TOPIC_AUTHORIZATION_FAILED) {
+                    unauthorizedTopics.add(metadata.topic());
```
Weren't we requesting update for this case earlier?

bdbyrne replied:
No, I believe because it's not considered a transient error.

bdbyrne replied:
So you were right to notice an issue. While we weren't requesting immediate update before, the original behavior still would have included this topic in the next update, as it should, since ACLs can change. This is fixed.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/MetadataCache.java:
```diff
@@ -105,6 +108,67 @@ Cluster cluster() {
+     * @param newClusterId the new cluster Id
+     * @param newNodes the new set of nodes
+     * @param addPartitions partitions to add
+     * @param addUnathorizedTopics unauthorized topics to add
```
nit: addUnathorizedTopics => addUnauthorizedTopics

bdbyrne replied:
Done.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerMetadata.java:
```diff
@@ -26,19 +26,40 @@
-    /* Topics with expiry time */
-    private final Map<String, Long> topics = new HashMap<>();
+    // The target size for a topic metadata refresh RPC.
+    static final int TOPIC_REFRESH_TARGET_SIZE = 20;
```
Was 20 chosen as the optimum size after testing?

bdbyrne replied:
Nope, by poll, as it's guaranteed to be no worse than it is today (i.e. all topics are urgent, which ignores this value). It actually doesn't come into play when there's large stress, it's more a background leveler, but there are cases that should be handled more gracefully, e.g. when there's a large number of topics or expiry time is short. I've updated the logic to handle this.

rajinisivaram reviewed (commented):
@bdbyrne Thanks for the PR, looks good. Left a few minor comments/questions.

rajinisivaram commented:
@bdbyrne The PR needs to be rebased. There were also a bunch of failures in the PR build that need looking into.

bdbyrne commented:
retest this please

bdbyrne commented:
One unforeseen issue is that some tests expect the Metadata to discover when topics have been deleted because they issue requests for all topic metadata and then rely on the cache to be rebuilt, thus discarding any topics that may have been deleted since the last update. With the merge code, the old metadata will be retained if it's still wanted, e.g. it matches the consumer's subscription pattern. Easy but non-elegant fix, looking for something cleaner.

bdbyrne commented:
retest this please

bdbyrne commented:
retest this please

bdbyrne commented:
retest this please

bdbyrne commented:
retest this please

bdbyrne commented:
Not sure if there's a limit on the number of iterations a PR can be tested, or what's going on, but I haven't been able to get a test run for a couple weeks now. Any ideas?

Commit:
KIP-526: Improve producer's metadata fetching for new topics.

When the producer encoutered new topic(s), it now only fetches the metadata for
the new topics. For cases where a producer interacts with a lot of topics, this
reduces the cost for the topic being evicted from the cache, and during startup
when populating the topic cache.

Additionally adds a new producer configuration variable 'metadata.max.idle.ms',
which controls how long topic metadata may be idle (i.e. not produced to) before
it's finally discarded from the metadata cache.

Commit:
(resolve conflicts)

bdbyrne commented:
retest this please

Commit:
(improve MetadataTest)

Commit:
(add new MetadataTest test)

ijuma commented:
ok to test

Commit:
(simplifications, fix test)

Commit:
(fix test due to last change)

bdbyrne commented:
retest this please

bdbyrne commented:
ok to test

rajinisivaram commented:
ok to test

rajinisivaram commented:
ok to test

hachikuji commented:
ok to test

hachikuji commented:
retest this please

ijuma commented:
retest this please

ijuma commented:
> Old: 1000 records sent, 32.446463 records/sec (0.00 MB/sec), 65.52 ms avg latency, 459.00 ms max latency, 23 ms 50th, 213 ms 95th, 360 ms 99th, 459 ms 99.9th.
> 
> New: 1000 records sent, 206.143063 records/sec (0.00 MB/sec), 10.55 ms avg latency, 187.00 ms max latency, 7 ms 50th, 29 ms 95th, 43 ms 99th, 187 ms 99.9th.

Awesome!

ijuma commented:
Btw, the 0.00 MB/s bit is wrong, right?

Commit:
When waiting, make sure a full refresh isn't requested.

bdbyrne commented:
> Btw, the 0.00 MB/s bit is wrong, right?

I was using tiny records since it's more about evaluating how fast the initial metadata is fetched. Both scenarios eventually have the metadata cached, and once the buffers start filling they're able to hit max throughput.

Old:
423 records sent, 84.1 records/sec (0.04 MB/sec), 22.0 ms avg latency, 157.0 ms max latency.
393 records sent, 77.9 records/sec (0.04 MB/sec), 27.2 ms avg latency, 119.0 ms max latency.
764 records sent, 152.0 records/sec (0.07 MB/sec), 21.6 ms avg latency, 140.0 ms max latency.
13077 records sent, 2615.4 records/sec (1.28 MB/sec), 19.1 ms avg latency, 259.0 ms max latency.
9681 records sent, 1936.2 records/sec (0.95 MB/sec), 7.2 ms avg latency, 19.0 ms max latency.
9091 records sent, 1818.2 records/sec (0.89 MB/sec), 8.0 ms avg latency, 27.0 ms max latency.
9947 records sent, 1989.0 records/sec (0.97 MB/sec), 8.6 ms avg latency, 25.0 ms max latency.
13094 records sent, 2617.8 records/sec (1.28 MB/sec), 10.0 ms avg latency, 35.0 ms max latency.
23459 records sent, 4690.9 records/sec (2.29 MB/sec), 12.2 ms avg latency, 31.0 ms max latency.
342941 records sent, 68588.2 records/sec (33.49 MB/sec), 248.4 ms avg latency, 1577.0 ms max latency.
952536 records sent, 190393.0 records/sec (92.97 MB/sec), 290.9 ms avg latency, 1866.0 ms max latency.
1008257 records sent, 201490.2 records/sec (98.38 MB/sec), 277.4 ms avg latency, 2268.0 ms max latency.
1008358 records sent, 201108.5 records/sec (98.20 MB/sec), 274.9 ms avg latency, 2075.0 ms max latency.
1093013 records sent, 218602.6 records/sec (106.74 MB/sec), 253.4 ms avg latency, 1681.0 ms max latency.

New:
8275 records sent, 1655.0 records/sec (0.81 MB/sec), 3.5 ms avg latency, 158.0 ms max latency.
12695 records sent, 2538.5 records/sec (1.24 MB/sec), 5.1 ms avg latency, 20.0 ms max latency.
9115 records sent, 1822.3 records/sec (0.89 MB/sec), 8.7 ms avg latency, 62.0 ms max latency.
10442 records sent, 2087.6 records/sec (1.02 MB/sec), 8.0 ms avg latency, 25.0 ms max latency.
13318 records sent, 2663.1 records/sec (1.30 MB/sec), 8.8 ms avg latency, 30.0 ms max latency.
21180 records sent, 4234.3 records/sec (2.07 MB/sec), 11.0 ms avg latency, 53.0 ms max latency.
255753 records sent, 51109.7 records/sec (24.96 MB/sec), 178.9 ms avg latency, 939.0 ms max latency.
1052303 records sent, 210250.3 records/sec (102.66 MB/sec), 265.4 ms avg latency, 1791.0 ms max latency.
1080308 records sent, 216018.4 records/sec (105.48 MB/sec), 257.4 ms avg latency, 1567.0 ms max latency.
1056188 records sent, 211153.1 records/sec (103.10 MB/sec), 259.6 ms avg latency, 1671.0 ms max latency.
1098043 records sent, 219608.6 records/sec (107.23 MB/sec), 254.9 ms avg latency, 1481.0 ms max latency.

I stress-tested a 10,000 single-partition topics setup to verify:

Old:
real	10m17.375s
user	7m31.150s
sys	0m2.766s

New:
real	0m40.362s
user	1m0.020s
sys	0m0.961s

The average latency was about 5-10x in the old, and max latency about 20-30x, which I wouldn't expect if it's just the metadata response payload size that's increased. In other words, there may be some improvements to be found on the broker, needs further profiling to see where the time is going.

Commit:
(another test)

Commit:
(two more higher-level expiry tests)

Commit:
(fix test race that could cause flakiness)

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/Metadata.java:
```diff
@@ -178,7 +187,7 @@ private synchronized boolean updateLastSeenEpoch(TopicPartition topicPartition,
             lastSeenLeaderEpochs.put(topicPartition, epoch);
             if (setRequestUpdateFlag) {
-                this.needUpdate = true;
+                this.needFullUpdate = true;
```
Why is this a full update?

bdbyrne replied:
In discussing the KIP with Colin and Jason, the common ground we found was to only perform partial metadata updates for new topics, and to leave all other cases as they currently are. So we don't track this particular topic and leave it as a full update.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/Metadata.java:
```diff
@@ -210,16 +219,19 @@ public synchronized boolean updateRequested() {
      */
-    public synchronized void update(MetadataResponse response, long now) {
-        this.update(this.requestVersion, response, now);
+    public synchronized void update(MetadataResponse response, long nowMs) {
```
This is a very confusing method. There is a risk that it may be invoked inadvertently when making changes, even though it is meant to be only for tests.

bdbyrne replied:
I'll change the name to something unwieldy, like updateWithCurrentRequestVersion(), and unify the two test-only functions.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/MetadataCache.java:
```diff
@@ -105,6 +109,109 @@ Cluster cluster() {
+        // a node has been modified or removed. Update the node object for known nodes, otherwise clear all nodes that were
+        // removed and let the metadata resolution process request an update, if necessary.
+        ArrayList<PartitionInfoAndEpoch> newPartitions = new ArrayList<>(addPartitions.size());
+        newPartitions.addAll(addPartitions);
```
Could just do `ArrayList<PartitionInfoAndEpoch> newPartitions = new ArrayList<>(addPartitions);`?

bdbyrne replied:
Done.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerMetadata.java:
```diff
@@ -101,8 +127,17 @@ public synchronized void awaitUpdate(final int lastVersion, final long timeoutMs
+        super.update(requestVersion, response, isPartialUpdate, nowMs);
+
+        // Remove all topics in the response that are in the new topic set. Note that if an error was encountered for a
+        // new topic's metadata, then any work to resolve the error will include the topic in a full metadata update.
```
So we don't retry using another partial request for a failed topic?

bdbyrne replied:
Correct, per discussion described above.

rajinisivaram replied:
If we see an authorization exception for example, would we retry using a full update straightaway or would it wait 5 minutes for the next refresh interval? Just want to make sure that the delay after these type of errors remains the same as before.

bdbyrne replied:
If a per-partition error is encountered, then the handling requests a full update, which will retry after the "retry.backoff.ms" has elapsed (default 100 ms).

bdbyrne replied:
Actually, authorization wouldn't retry right away. Errors deriving from InvalidMetadataException would, however. Handling of both cases is left unchanged.

rajinisivaram commented on clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java:
```diff
@@ -1037,7 +1126,7 @@ public void testCloseIsForcedOnPendingAddOffsetRequest() throws InterruptedExcep
         Time time = new MockTime();
         MetadataResponse initialUpdateResponse = TestUtils.metadataUpdateWith(1, singletonMap("testTopic", 1));
-        ProducerMetadata metadata = newMetadata(0, Long.MAX_VALUE);
+        ProducerMetadata metadata = newMetadata(0, Long.MAX_VALUE, Long.MAX_VALUE);
```
Is there a reason why every test seems to be using the same expiration and idle time? Earlier, we would have used default idle time?

bdbyrne replied:
No reason. I've reverted the functionality to be as it was using 5 minutes for the test.

rajinisivaram reviewed (commented):
@bdbyrne Thanks for the PR, looks good. Left a few minor comments. It may be useful to run one of the integration tests with very low `metadata.max.idle.ms` to ensure we trigger this code path in integration tests. Have we checked with @mumrah that it is ok to merge this into 2.5 branch?

rajinisivaram commented:
Looks like this PR needs rebasing.

Commit:
(address review comments, fix merge conflict)

bdbyrne commented:
> Looks like this PR needs rebasing.

Done. New metadata handling in MetadataCache simplified the node resolution code for me.

rajinisivaram commented:
ok to test

rajinisivaram reviewed (approved):
@bdbyrne Thanks for the updates, LGTM

rajinisivaram commented:
@bdbyrne Thanks for the PR, merging to trunk.
